accepted/future-releases/0546-patterns/exhaustiveness.md

```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
-handling for integers that won't handle cases like:
+7.  Return the result of is-exhaustive on `value` and `remaining`.
 
-```dart
-switch (list) {
-  case [var a, var b]: ...
-  case [var a, var b]: ...
-}
-```
+    *Now that we've removed any cases that might not match based on the current
+    space we're looking at, we can proceed forward along the worklists.*
 
-The algorithm currently won't detect that the second space is unreachable. We
-will probably want to support `...` in list patterns to allow matching on lists
-of unknown length where the length is at least above some minimum, as in:
+### Unpacking properties
 
-```dart
-switch (list) {
-  case [var a, ..., var b]: ...
-  case [var a, var b]: ...
-}
-```
+A key challenge with exhaustiveness checking is that patterns are arbitrarily
+deeply nested trees. At the same time, we need to consider a set of cases in
+parallel to see if they cover a value. And those cases may have patterns with
+different sets of properties and different amounts of nested subpatterns.
 
-In that case, to do smart exhaustiveness checks, we probably want the algorithm
-to understand integer ranges. Most other languages do this, so it should be
-tractable.
+To handle that, the algorithm does an incremental depth-first traversal of the
+value and case spaces in parallel. After applying any filtering from a space's
+type and restriction, we can discard the space itself, but we need to recurse
+into its properties too.
 
-### Shared subtypes of sealed types
+We do that here. Unpacking a space's properties prepends them to the worklist so
+that the depth-first traversal will explore into them. We use `keys` and pad the
+worklist with match-all spaces in order to keep all of the worklists aligned.
 
-The expand type procedure assumes that any type is a direct subtype of at most
-*one* sealed supertype. (It can have other supertypes but only one *sealed*
-one.) It assumes that the language forbids a class hierarchy like:
+To unpack a `space` given a set of `keys`:
 
-```
-   (A)
-   / \
- (B) (C)
- / \ / \
-D   E   F
-```
+1.  Let `result` be an empty list of space unions.
 
-Here, `A`, `B`, and `C` are sealed supertypes. The subtypes of `A` are `B` and
-`C`. The subtypes of `B` are `D` and `E`. The subtypes of `C` are `E` and `F`.
-So `E` is a direct subtype of both `B` and `C`.
+2.  For each key in `keys`:
 
-If expanding supported class hierarchies like this correctly then given a switch
-like this:
+    1.  If `space` has a property with this key, then append that property's
+        subspace union to `result`.
 
-```dart
-A a = ...
-switch (a) {
-  case B b: ...
-  case F f: ...
-}
-```
+    2.  Else, append a `space` to `result` whose type is `Object?`. *If a
+        pattern doesn't have a subpattern for some field or getter, then it
+        matches all values that the field or getter could return, which is
+        equivalent to it matching it with type `Object?`.*
```

Comment:
While this is true for the case spaces it isn't precise enough for the value space since it might lead to a space larger that what the type of the property can contain. For instance, if the value space is `Object()` and we have a case space `Object(hashCode: int())` then unpacking the key `hashCode` for `Object()` to `Object?()` would mean that we would conclude that `int()` does not exhaust the `hashCode` property. Instead we need to use the type corresponding to the property of the type of the space when it exists. Currently the algorithm uses `Object?()` as a fallback when the property doesn't exist, which (I think) can only happen for erroneous code. (Given that we have checked that the type of the value space is a subtype of the case space, so if the property exists on the case space type it must exist on the value space type.)

Reply:
Ah, you're right. I felt like I was getting something wrong here. Fixed. (And added a section up on properties defining how to get the type of a key.)